nixos/modules/programs/fish.nix

```diff
@@ -169,6 +169,71 @@ in
+        # joins with null byte to acommodate all characters in paths, then respectively gets all paths before / after the first one including "generated_completions",
+        # splits by null byte, and then removes all empty lines produced by using 'string'
+        set -l prev (string join0 $fish_complete_path | string match --regex "^.*?(?=\x00[^\x00]*generated_completions.*)" | string split0 | string match -er ".")
+        set -l post (string join0 $fish_complete_path | string match --regex "[^\x00]*generated_completions.*" | string split0 | string match -er ".")
+        set fish_complete_path $prev "/etc/fish/generated_completions" $post
+      end
+    '';
+
+    environment.etc."fish/generated_completions".source =
+      let
+        patchedGenerator = pkgs.stdenv.mkDerivation {
+          name = "fish_patched-completion-generator";
+          srcs = [
+            "${pkgs.fish}/share/fish/tools/create_manpage_completions.py"
+            "${pkgs.fish}/share/fish/tools/deroff.py"
+          ];
+          unpackCmd = "cp $curSrc $(basename $curSrc)";
+          sourceRoot = ".";
+          patches = [ (pkgs.writeText "fish_completion-generator_patch" ''
+            --- a/create_manpage_completions.py
+            +++ b/create_manpage_completions.py
+            @@ -776,8 +776,6 @@ def parse_manpage_at_path(manpage_path, output_directory):
+
+                         built_command_output.insert(0, "# " + CMDNAME)
+
+            -            # Output the magic word Autogenerated so we can tell if we can overwrite this
+            -            built_command_output.insert(1, "# Autogenerated from man page " + manpage_path)
+                         # built_command_output.insert(2, "# using " + parser.__class__.__name__) # XXX MISATTRIBUTES THE CULPABILE PARSER! Was really using Type2 but reporting TypeDeroffManParser
+
+                         for line in built_command_output:
+          '') ];
+          dontBuild = true;
+          installPhase = ''
+            mkdir -p $out
+            cp * $out/
+          '';
+        };
+        generateCompletions = package: pkgs.runCommand
+          "${package.name}_fish-completions"
+          (
+            {
+              inherit package;
+              nativeBuildInputs = [ pkgs.python3 ];
+              buildInputs = [ pkgs.fish ];
+              preferLocalBuild = true;
+              allowSubstitutes = false;
+            }
+            // optionalAttrs (package ? meta.priority) { meta.priority = package.meta.priority; }
+          )
+          ''
+            mkdir -p $out
+            if [ -d $package/share/man ]; then
+              find $package/share/man -type f | xargs python ${patchedGenerator}/create_manpage_completions.py --directory $out >/dev/null
+            fi
+          '';
+      in
+        pkgs.buildEnv {
+          name = "system_fish-completions";
+          ignoreCollisions = true;
+          paths = map generateCompletions config.environment.systemPackages;
```

Comment:
Is there a way to make this work not only for the packages in `systemPackages` but also for `programs.*.enable = true`?

Reply:
Somehow accessing what will later become `sw/share/man/` would probably lead to an infinite recursion, but maybe someone has an idea.

Reply:
Every module I've seen so far uses systemPackages to install their packages, are there any that use a different mechanism?

Reply:
You're right, they do :D Never mind.